File: FreeCodeCamp/Data Structures/Circular Queue/circularQueue.py
class CircularQueue:
    def __init__(self, size):
        self.queue = []
        self.read = 0
        self.write = 0
        self.max = size - 1
        
        while (size > 0):
            self.queue.append(None)
            size -= 1
    
    def enqueue(self, item): 
        if self.write >= self.max + 1 + self.read:
            return None
        self.queue[self.write % (self.max + 1)] = item
        self.write += 1
        return item
        
    def dequeue(self):
        if self.read >= self.write:
            return None
        item = self.queue[self.read % (self.max + 1)]
        self.queue[self.read % (self.max + 1)] = None
        self.read += 1
        return item

File: FreeCodeCamp/Data Structures/Circular Queue/circularQueue.js
class CircularQueue {
  constructor(size) {

    // Initialization 
    this.queue = [];
    this.read = 0;
    this.write = 0;
    this.max = size - 1;

    while (size > 0) {
      this.queue.push(null);
      size--;
    }
  }

  print() {
    return this.queue;
  }

  enqueue(item) {

    if (this.write >= this.max + 1 + this.read) { // Check if the queue is full 
      return null
    } 

    this.queue[this.write % (this.max + 1)] = item
    this.write ++
    return item

  }

  dequeue() {

    if (this.read >= this.write) { // Check if the queue is empty; 
      return null
    };
       const item = this.queue[this.read % (this.max + 1)]
       this.queue[this.read % (this.max + 1)] = null
       this.read = this.read + 1
       return item // Item is read - 1 positioned now
    }
  }

let queue = new CircularQueue(5)

queue.enqueue('A')
queue.enqueue('B')
queue.enqueue('C')
queue.enqueue('D')
queue.enqueue('E')

queue.dequeue()
